### PR TITLE
Move ConstantFun to akka.util

### PR DESCRIPTION
**Pull request by Hawstein**
Refs #22813

Also:
* Change some un-exposed variables to `private`
* Include the `nullFun` which is used in #22801

### CODE REVIEW

File: akka-actor/src/main/scala/akka/util/ConstantFun.scala [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /**
  * Copyright (C) 2015-2017 Lightbend Inc. <http://www.lightbend.com>
  */
-package akka.stream.impl
+package akka.util
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ConstantFun.scala]: I was thinking we perhaps should still just copy it and mark as deprecated.
The reason is that Akka HTTP did use those (!), so if we remove them, and someone upgrades akka but not Akka HTTP they would break.

I propose to keep it deprecated for one or two HTTP releases so we can move to touching the new one in akka.util.

**Review by ktoso — CHANGES_REQUESTED** [review]
I think we do have to copy it for now, otherwise looks good!

File: akka-actor/src/main/scala/akka/util/ConstantFun.scala (hunk 1 above)
Comment (johanandren) [inline akka-actor/src/main/scala/akka/util/ConstantFun.scala]: Is it really worth breaking comp over this at all, having them in two places (one deprecated) is two allocs instead of one, hardly something that will swamp the JVM...

Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ConstantFun.scala]: Then I'd still propose to copy it over and deprecate old ones.
It's very noticeable changes in quite a few hotpaths and having a place for them is a nice thing to have.

**Comment by Hawstein** [comment]
Oops，forgot we have a separate `akka-http` project...

**Comment by Hawstein** [comment]
PR is updated. @ktoso @johanandren

File: akka-actor/src/main/scala/akka/util/ConstantFun.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import akka.annotation.InternalApi
+import akka.japi.function.{ Function ⇒ JFun, Function2 ⇒ JFun2 }
+import akka.japi.{ Pair ⇒ JPair }
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object ConstantFun {
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/util/ConstantFun.scala]: @ktoso if Akka HTTP uses it, shouldn't it be public API then?

Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ConstantFun.scala]: Hm, good point - OR Akka HTTP should not touch it and use it's own place for those would be even better I think.

The reason it still used those is historical I guess, since it was together with streams, it had access there, and then we just "moved" the classes to new repo while it's access privileges to this class remained...

Comment (Hawstein) [inline akka-actor/src/main/scala/akka/util/ConstantFun.scala]: hmm, indeed, Akka HTTP use it for some historical reasons. I think here we can just move `ConstantFun` without worrying about Akka HTTP using it because it uses a specific version of Akka. The end users  upgrading Akka but not Akka HTTP will break nothing. The only thing we should care is that when upgrading Akka for Akka HTTP, we should change the usage of `ConstantFun`, or maybe use its own copy in Akka HTTP project.

Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ConstantFun.scala]: That's not the right conclusion, people may want to just upgrade an Akka version in their Akka HTTP 10.0.6, for example if Akka has a critical bugfix they should be able to just upgrade that. Thus, I don't think we can just move it - it has to be a careful move away (copy the class, deprecate it etc).

Comment (Hawstein) [inline akka-actor/src/main/scala/akka/util/ConstantFun.scala]: Hm, I guess few people will play around their own version of Akka HTTP and even deploy their modified version on production. However, maybe you are right, we should take care of that situation.

Since Akka HTTP was just moved from Akka to a new repo, I think there might be more *INTERNAL APIs* in Akka are touched by Akka HTTP. I just did a very quick check, `OptionVal` in `akka.util` is also used in several places in Akka HTTP. If Akka HTTP should use their own version of those internal stuffs, should we create another issue to fix it together?

Comment (hepin1989) [inline akka-actor/src/main/scala/akka/util/ConstantFun.scala]: Vote for copy and pasting.
For the critical bugfix, that maybe only for commercial users:)

Comment (Hawstein) [inline akka-actor/src/main/scala/akka/util/ConstantFun.scala]: @hepin1989 Make sense!

File: project/MiMa.scala [hunk 3]
```diff
@@ -1166,27 +1166,31 @@ object MiMa extends AutoPlugin {
...
+        ProblemFilters.exclude[MissingClassProblem]("akka.stream.impl.fusing.GroupedWithin"),
+
+        ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.stream.Graph.traversalBuilder"),
+        ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.stream.Graph.named"),
+        ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.stream.Graph.addAttributes"),
+        ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.stream.Graph.async"),
+
+        // #22818 Move ConstantFun to akka.util
+        ProblemFilters.exclude[MissingClassProblem]("akka.stream.impl.ConstantFun"),
+        ProblemFilters.exclude[MissingClassProblem]("akka.stream.impl.ConstantFun$")
```
Comment (patriknw) [inline project/MiMa.scala]: these new should now be in the 2.5.1 section, please rebase

Comment (patriknw) [inline project/MiMa.scala]: but it wasn't removed so these should be removed, just revert the changes of this file (skip the whitespace cleanup)

**Review by patriknw — CHANGES_REQUESTED** [review]
looking good, change mima

**Comment by Hawstein** [comment]
@patriknw The PR is updated.

**Review by johanandren — APPROVED** [review]
LGTM

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by ktoso** [comment]
Thanks a lot @Hawstein !
